[IMAGING-345] Unit test : make tests in memory instead of writing hundred of tmp files

kpouer opened the pull request:
An attempt to perform unit tests in memory using byte array when possible instead of writing temporary files.
Tests pass much faster and do not make useless write on my SSD.
I know that having the tmp files might be useful in some case, if you  think it it necessary I could update the patch to write tmp file in case of failed test.

Commit:
[IMAGING-345] perform tests in memory instead of disk making them faster

kinow reviewed (commented):
I'm busy with $work these days, but

>An attempt to perform unit tests in memory using byte array when possible instead of writing temporary files.
>Tests pass much faster and do not make useless write on my SSD.
>I know that having the tmp files might be useful in some case, if you think it it necessary I could update the patch to write tmp file in case of failed test.

I think it could be useful in some cases, like if you want to debug/troubleshoot it, but I think speeding up the tests and reducing IO sound like a good idea. In case a test fails and we need to troubleshoot, it shouldn't be too hard to write it to disk to investigate (assuming that's needed at all). So I'm a big +1 for this, @kpouer .

I had a quick look and looks like you are adding some public methods too, in non-test files. Were these necessary for the change in this PR, @kpouer ?

Final note, it'd be necessary to squash the commits later. If you are still sorting out how to organize the change, feel free to keep it as is until you have it ready. Otherwise, squash it so that it's ready to be merged, please.

Thanks heaps!

garydgregory commented:
Keep in mind that tests must be able to run in containers that may not provide much memory.

kpouer commented:
Hi @kinow,
About the new methods, it is utility that get various information about the image. Some format are not detected only from their data and the file extension is necessary. It was of course working with tmp files, but not with byte arrays.
I took inspiration of the equivalent method taking an InputStream and a filename, for example
There was a 

public static ICC_Profile getICCProfile(final InputStream is, final String fileName)

Then I created
public static ICC_Profile getICCProfile(final String fileName, final byte[] bytes) (the parameter order is not the same but it follows the constructor parameter orders of the ByteSourceArray class (however when rereading I think that Imaging class should always have the same parameter orders, it would be easier to understand.
I could have used getICCProfile(ByteSource) but the method was protected then I should have changed it's visibility.

About the troubleshooting I'll update the branch in order to save tmp files for failed test.

@garydgregory I don't think it will take a lot of memory as images are processed one by one, I think it should store something like 2 or 3 time the same image during some tests, then the memory is released to the GC to process the next image. And test images do not seem to be huge. However I am interested but not sure about what you mean by tested in containers. Do I have a way to test it to validate ?

garydgregory commented:
I was referring to GitHub Actions builds being resource constrained (to what extent I don't know exactly).

kpouer commented:
Hey, I was able to do the same test without changing public api so I reverted the change (even if I think those functions might be useful, it could be done in another PR if necessary.
I added code to save tmp file in case of error during test (catched throwable, I don't like it but had no idea how to do it in another way). For memory I tried to isolate ByteArrayOutputStream in blocks to allow GC to collect them as soon as possible, I hope it helps.
when you think it is ok I can resubmit the same patch in only one commit (or update it if you think there are more changes to do

kinow commented:
Thanks @kpouer can you squash your commits? And would you be able to create a JIRA for this? Even though it's a change in the tests I'd like to review (will try to compare number of IO ops, and memory on my laptop at least) and include it in the changelog for the next release. Thanks!

kinow reviewed (commented):
Thanks @kpouer can you squash your commits? And would you be able to create a JIRA for this? Even though it's a change in the tests I'd like to review (will try to compare number of IO ops, and memory on my laptop at least) and include it in the changelog for the next release. Thanks!

kpouer commented:
Hey, I was not sure how I could squash commits without deleting my branch and pushing it again so I created a new branch and a new pull request (same code but squashed into one commit)
https://github.com/apache/commons-imaging/pull/272
I would be happy to create a JIRA ticket but I don't have an account on it. It says I could send an email to private@commons-imaging.apache.org if that email is setup. Is it ?

kinow commented:
Hi @kpouer 

> Hey, I was not sure how I could squash commits without deleting my branch and pushing it again so I created a new branch and a new pull request (same code but squashed into one commit) #272

It's probably useful to get used to squashing commits, both for Apache Commons code contributions, but for other OSS that follow similar approach. The issue with opening new PR's is that a) you may have to add more commits again, and then squash again, and b) it gets more complicated to keep track of discussions and reviews (there's also c) where users/devs subscribed to the repo may get spammed by similar changes).

If you'd like you can try on this PR, as it's still open: https://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

I squash rebasing interactively. Basically what I'd do here would be something similar to:

```bash
git fetch --all
git rebase upstream/master # upstream for me is apache/commons-imaging, kinow for me is the kinow/commons-imaging
gitk # or git log -n N, but I like to take a look at the GUI too, then count the number of commits since the top of master
# you have 17 commits here, ATOW
git rebase -i HEAD~17  # -i is interactive, i.e. I will interact with git to decide how to rebase
# each commit is prefixed by an action code, I mark the first line as r-eword
# and for each new commit line in the VIM/editor, I mark as f-ixup
```

![image](https://user-images.githubusercontent.com/304786/213907127-bdb2520e-d12c-4ab7-b87f-c5171c9069b9.png)

When I save my editor, `git rebase` proceeds, opening the editor again as I asked to reword the fix commit. Now I write a good commit line. If I already have the JIRA number I prefix the commit message with [IMAGING-NNNN] Perform tests in memory instead of disk, making them faster".

When I save this file again, git will use the text as the commit message of the reworded commit. The other commits have been all "fixed-up". So now everything has been "squashed" into a single commit.

At this point I `gitk` again, or `git log -n 1` or `-n 2` to look at the commit history again.

![image](https://user-images.githubusercontent.com/304786/213907214-c6522b8a-4582-4f46-b50f-294bc307d108.png)

At this point your local branch contains the squashed commits for your change. You can browse the code, run `mvn`, etc., until you are sure that everything is in order. Then you are ready to `push -f kinow/branch-name` (where kinow will be your personal remote, and branch-name must be `unit_test_tmp_fil....`, I think).

And the pull request will have been updated with your forced (-f) commit, showing a single commit for reviewers :+1: 

If that works you can close your other PR. Or if you prefer to try this later, feel free to close this issue :+1: up to you.

> I would be happy to create a JIRA ticket but I don't have an account on it. It says I could send an email to [private@commons-imaging.apache.org](mailto:private@commons-imaging.apache.org) if that email is setup. Is it ?

I am sorry, due to bots SPAMing our JIRA server we had to lock it for new users :disappointed_relieved:

Send the email, please, to the commons private mail (it's the email address you mentioned above, minus the "-imaging" part) with the info on https://infra.apache.org/jira-guidelines.html#account, and I or another committer will create your account.

Really sorry for the inconvenience :bow: 

Thanks @kpouer !

kinow commented:
>  kpouer force-pushed the unit_test_tmp_file_optimization branch from eddd5bf to 1947c86 

:clap: great job! Much easier to review, and thanks for closing the other PR. Bravo @kpouer :+1:

kpouer commented:
Yes I was able to squash but was missing the "force" option for the push

kinow commented:
Sorry for the delay, next on my list! :nerd_face:

kinow commented on src/test/java/org/apache/commons/imaging/roundtrip/RoundtripBase.java:
```diff
@@ -39,35 +38,52 @@ public class RoundtripBase {
         final ImagingParameters params = Util.getImageParser(formatInfo.format).getDefaultParameters();
-        try (FileOutputStream fos = new FileOutputStream(temp1)) {
-            imageParser.writeImage(testImage, fos, params);
+        try (ByteArrayOutputStream bos = new ByteArrayOutputStream(1000000)) {
```
Maybe leave it the default value and let it handle expanding the internal buffer? That could be helpful if we load many smaller images, I think (although I have no idea the distribution of sizes of our images & their metadata too).

kinow commented on src/test/java/org/apache/commons/imaging/formats/bmp/BmpRoundtripTest.java:
```diff
@@ -140,17 +139,21 @@ private void writeAndReadImageData(final int[][] rawData) throws IOException, Im
+        } catch (final Throwable e) {
+            final Path tempFile = Files.createTempFile("temp", ".bmp");
+            Files.write(tempFile, bytes);
+            System.err.println("Failed tempFile " + tempFile);
```
Also, reading again the code, I think we can rethink this fallback mechanism.

Right now, in case there's any exception of any type, it writes the image to the disk. But that might end up doing that unnecessarily for some exception types.

It might be simple to just throw the error and let a developer call the code to write it to the disk, I think. Otherwise we will also have to remember to use this boilerplace code whenever we write new test code that loads images into the memory.

Sorry @kpouer, as I believe your initial PR didn't have this, and you added it based on our initial feedback. Right now I'm included to undo that, but give me some time to think about this one :+1:

kinow replied:
(and no need to worry, I can undo the change if we decide to go down that path :wink: :+1: )

kpouer replied:
Hi, in fact I could undo the change that automatically write the tmp file as it is probably useless in case of automatic tests. But add back the code that writes the file but commented out so it could be activated manually by a developper if needed ?

kinow replied:
>Hi, in fact I could undo the change that automatically write the tmp file as it is probably useless in case of automatic tests. But add back the code that writes the file but commented out so it could be activated manually by a developper if needed ?

That'd be useful! I really think that'd be better than just writing whenever there's an exception. Sorry for the extra work @kpouer !

kinow commented:
Used Docker to run the build from this PR limiting the allowed memory for the container. The build crashed at 64mb, 128mb, 256mb. And it passed for 512mb and higher.

Switched to `master`, and 512mb passes, and 256mb fails as well. Command line used:

```bash
docker run -it --rm --name my-maven-project -v "$(pwd)":/usr/src/mymaven -w /usr/src/mymaven --memory=256m maven:3.8-openjdk-17 mvn clean install
```

kpouer commented:
I updated the branch, first I wanted to comment out code to create tmp files for troubleshooting but finally I just did enclose it in a if (false) {} block, this way no problem with imports and it is even easiest to set it to true for troubleshooting.
I'm exploring buiding with docker just to see if something can be done more efficiently

kinow commented on src/test/java/org/apache/commons/imaging/formats/png/PngTextTest.java:
```diff
@@ -74,8 +76,10 @@ public void test() throws Exception {
-        FileUtils.writeByteArrayToFile(tempFile, bytes);
+        if (false) { // activate saving tmp file for troubleshooting
+            final File tempFile = Files.createTempFile("temp", ".png").toFile();
+            FileUtils.writeByteArrayToFile(tempFile, bytes);
```
`LOGGER` no used here.

kinow commented on src/test/java/org/apache/commons/imaging/formats/png/PngWriteForceTrueColorText.java:
```diff
@@ -21,21 +21,25 @@
         final List<File> images = getPngImages();
         for (final File imageFile : images) {
 
+            byte[] outFile = null;
```
This could be inside the `try` block. I thought it was declared outside for some reason, but apparently it's only used within that block?

kinow commented on src/test/java/org/apache/commons/imaging/formats/png/PngWriteForceTrueColorText.java:
```diff
@@ -50,13 +54,19 @@ public void test() throws Exception {
+
+                if (false) { // activate saving tmp file for troubleshooting
+                    final Path tempFile = Files.createTempFile(imageFile.getName() + ".", ".jpg");
+                    Files.write(tempFile, outFile);
```
`LOGGER` not used here.

kinow commented on src/test/java/org/apache/commons/imaging/roundtrip/RoundtripBase.java:
```diff
@@ -62,10 +78,18 @@ protected void roundtrip(final FormatInfo formatInfo, final BufferedImage testIm
+                final File temp2 = Files.createTempFile(tempPrefix + ".", "." + formatInfo.format.getDefaultExtension()).toFile();
+                try (FileOutputStream fos = new FileOutputStream(temp2)) {
+                    imageParser.writeImage(image2, fos, params);
+                }
```
`LOGGER` not used here.

kinow reviewed (commented):
@kpouer I'd be OK removing the `if (false)` block, or leaving it commented out in one test class, or adding a `README.md` in the test directory, etc.

As more tests are added, I guess other devs may ignore copying that `if (false)` block, and it's not very hard to change the code to write to the disk in case a test is failing.

Furthermore, it's hard to be consistent with this code as it's not used (see comments where in some blocks there's a call to `LOGGER.severe`, while others are missing it.

What do you think?

kpouer commented:
Hi @kinow I think you are right, those blocks adds complexity to the test and I am not sure they will be so useful. In the rare case someone really needs to see the file it is finally easy to add a few lines of code to write the file. I will remove them.

kinow commented:
> Hi @kinow I think you are right, those blocks adds complexity to the test and I am not sure they will be so useful. In the rare case someone really needs to see the file it is finally easy to add a few lines of code to write the file. I will remove them.

Excellent @kpouer . Feel free to ignore/resolve my last review comments then. I believe most/all are about the missing `LOGGER` in the `if(false)` block, and so they shouldn't be valid any longer. Thank you!

kinow commented on src/test/java/org/apache/commons/imaging/roundtrip/NullParametersRoundtripTest.java:
```diff
@@ -38,13 +38,17 @@ public static Stream<FormatInfo> data() {
+        Imaging.getImageInfo(new ByteArrayInputStream(temp1), filename);
+        Imaging.getImageSize(new ByteArrayInputStream(temp1), filename);
+        Imaging.getMetadata(new ByteArrayInputStream(temp1), filename);
+        Imaging.getICCProfile(new ByteArrayInputStream(temp1), filename);
```
Hmmm, we cannot resuse the same `ByteArrayInputStream` right? :thinking: would be nicer if we closed these streams at least. Maybe we have to split it into separated test methods?

kpouer replied:
In fact we can reuse the same ByteArrayInputStream but the code is more verbose, and closing it has no effect (however it is useful for static analysis of IDE so they do not complains about it ;) )
try (ByteArrayInputStream bos = new ByteArrayInputStream(temp1)) {
            Imaging.getImageInfo(bos, filename);
            bos.reset();
            Imaging.getImageSize(bos, filename);
            bos.reset();
            Imaging.getMetadata(bos, filename);
            bos.reset();
            Imaging.getICCProfile(bos, filename);
            bos.reset();
            final BufferedImage imageRead = Imaging.getBufferedImage(bos, filename);
            bos.reset();
            assertNotNull(imageRead);
        }

Another option is to use the methods taking directly a byte[] but it would require to create new methods also taking the filename as the file extension is sometimes used to help reading data (unfortunately some exists but the parameter order is not consistent, sometimes the filename comes first some other time it comes after)

kinow commented on src/test/java/org/apache/commons/imaging/roundtrip/ImageAsserts.java:
```diff
@@ -99,4 +99,18 @@ static void assertEquals(final File a, final File b) throws IOException {
+            }
+            Assertions.assertEquals(aByte, bByte);
+        }
+    }
```
Is there a reason for using this over using JUnit's [`assertArrayEquals`](https://junit.org/junit5/docs/5.0.0/api/org/junit/jupiter/api/Assertions.html#assertArrayEquals-byte:A-byte:A-) ? The `Debug` is not active by default, I believe, so using this doesn't seem to add must value.

kinow replied:
I think the JUnit version should work too. Not sure if what `ImageAsserts#assertEquals(File, File)` is doing is necessary. It's reading the bytes from the stream, I assume, without handling `int` values (and thus having to apply a filter mask `0xff` to clear the other bits in the 32 byte-Java).

The point here is that if the JUnit's `assertArrayEquals` work, it's better it's less code for us to test :+1:

kinow commented on src/test/java/org/apache/commons/imaging/roundtrip/FormatInfo.java:
```diff
@@ -130,4 +130,16 @@ class FormatInfo {
+            ", colorSupport=" + colorSupport +
+            ", identicalSecondWrite=" + identicalSecondWrite +
+            ", preservesResolution=" + preservesResolution +
+            '}';
```
Looks harmless, but was there a reason for adding this one?

kpouer replied:
It helped me a little when looking at some FormatInfo in the debugger of my IDE but if you don't find it useful I can remove that.

kinow replied:
Got it. Sounds like something that could be helpful to others too, and tests only, so really no harm in adding it. Thanks for the explanation!

kinow reviewed (changes requested):
See my comment about the JUnit method, but I think this is my last review feedback. It should be ready for a final pass of :eyes: and merge. Thanks @kpouer !!!

kpouer commented:
Hey I missed the notification, yes a simple assertArrayEquals would be more efficient, do you want that I remove it ?

kinow commented:
>Hey I missed the notification, yes a simple assertArrayEquals would be more efficient, do you want that I remove it ?

Yes, if possible. Rebase and fix the conflicts, and then we can get it ready to be merged. Thanks @kpouer !

kpouer commented:
done

kinow reviewed (approved):
Excellent, @kpouer ! I will add a changelog and merge it. Thank you for your contribution, and for the patience!

kinow commented:
Merged!
